src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -0,0 +1,114 @@
+from pip._vendor.packaging.utils import canonicalize_name
+
+from pip._internal.req.constructors import install_req_from_line
+from pip._internal.req.req_install import InstallRequirement
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+from .base import Candidate
+
+if MYPY_CHECK_RUNNING:
+    from typing import Dict, Optional, Sequence
+
+    from pip._internal.models.link import Link
+    from pip._internal.operations.prepare import RequirementPreparer
+    from pip._internal.resolution.base import InstallRequirementProvider
+
+    from pip._vendor.packaging.version import _BaseVersion
+    from pip._vendor.pkg_resources import Distribution
+
+
+# Dummy to make lint pass
+_CANDIDATE_CACHE = {}  # type: Dict[Link, Candidate]
+
+
+def make_candidate(
+    link,             # type: Link
+    preparer,         # type: RequirementPreparer
+    parent,           # type: InstallRequirement
+    make_install_req  # type: InstallRequirementProvider
+):
+    # type: (...) -> Candidate
+    if link not in _CANDIDATE_CACHE:
+        _CANDIDATE_CACHE[link] = LinkCandidate(
+            link,
+            preparer,
+            parent=parent,
+            make_install_req=make_install_req
+        )
+    return _CANDIDATE_CACHE[link]
+
+
+def make_install_req_from_link(link, parent):
+    # type: (Link, InstallRequirement) -> InstallRequirement
+    # TODO: Do we need to support editables?
+    return install_req_from_line(
+        link.url,
+        comes_from=parent.comes_from,
+        use_pep517=parent.use_pep517,
+        isolated=parent.isolated,
+        wheel_cache=parent._wheel_cache,
+        constraint=parent.constraint,
+        options=dict(
+            install_options=parent.install_options,
+            global_options=parent.global_options,
+            hashes=parent.hash_options
+        ),
```

Comment:
Why `dict()` instead of `{...}`?

Reply:
Because I don't like having to quote the keys 😃 Just a personal preference, though.